models/pulls/pull_review.go

```diff
@@ -0,0 +1,109 @@
+	review := &PRReview{UserID: userID, CommitSHA: commitSHA, PullID: pullID}
+	has, err := db.GetEngine(db.DefaultContext).Get(review)
+	return review, has, err
+}
+
+// UpdateReview updates the given review inside the database, regardless of whether it existed before or not
+// The given map of viewed files will be merged with the previous review, if present
+func UpdateReview(userID, pullID int64, commitSHA string, viewedFiles map[string]bool) error {
+	review, exists, err := GetReview(userID, pullID, commitSHA)
+	if err != nil {
+		return err
+	}
+
+	if exists {
+		review.ViewedFiles = mergeFiles(review.ViewedFiles, viewedFiles)
+	} else if previousReview, err := getNewestReviewApartFrom(commitSHA, userID, pullID); err != nil {
+		return err
+
+		// Overwrite the viewed files of the previous review if present
+	} else if previousReview != nil {
+		review.ViewedFiles = mergeFiles(previousReview.ViewedFiles, viewedFiles)
+	} else {
+		review.ViewedFiles = viewedFiles
+	}
+
+	// Insert or Update review
+	engine := db.GetEngine(db.DefaultContext)
+	if !exists {
+		_, err := engine.Insert(review)
+		return err
+	}
+	_, err = engine.ID(review.ID).Update(review)
+	return err
+}
+
+// mergeFiles merges the given maps of files with their viewing state into one map.
+// Values from oldFiles will be overridden with values from newFiles
+func mergeFiles(oldFiles, newFiles map[string]bool) map[string]bool {
+	if oldFiles == nil {
+		return newFiles
+	} else if newFiles == nil {
+		return oldFiles
+	}
+
+	for file, viewed := range newFiles {
+		oldFiles[file] = viewed
+	}
+	return oldFiles
+}
+
+// GetNewestReview gets the newest review of the current user in the current PR.
+// The returned PR Review will be nil if the user has not yet reviewed this PR.
+func GetNewestReview(userID, pullID int64) (*PRReview, error) {
+	var review PRReview
+	has, err := db.GetEngine(db.DefaultContext).Where("user_id = ?", userID).And("pull_id = ?", pullID).OrderBy("updated_unix DESC").Limit(1).Get(&review)
+	if err != nil || !has {
+		return nil, err
+	}
+	return &review, err
+}
+
+// getNewestReviewApartFrom is like GetNewestReview, except that the second newest review will be returned if the newest review points at the given commit.
+// The returned PR Review will be nil if the user has not yet reviewed this PR.
+func getNewestReviewApartFrom(commitSHA string, userID, pullID int64) (*PRReview, error) {
+	var reviews []PRReview
+	err := db.GetEngine(db.DefaultContext).Where("user_id = ?", userID).And("pull_id = ?", pullID).OrderBy("updated_unix DESC").Limit(2).Find(&reviews)
+	// It would also be possible to use ".And("commit_sha != ?", commitSHA)" instead of the error handling below
+	// However, benchmarks show a MASSIVE performance gain by not doing that: 1000 ms => <300 ms
+
+	// Error cases in which no review should be returned
+	if err != nil || len(reviews) == 0 || (len(reviews) == 1 && reviews[0].CommitSHA == commitSHA) {
+		return nil, err
```

Comment:
I don't know how to avoid that.
Unlike in Java, I don't have an `Optional` at hand I could simply return.
The best I can do is simply notify about it in the documentation and hope that whoever calls the method reads the documentation beforehand.